File: packages/py/src/qorechain/messages/registry.py
"""A type-URL → protobuf-class registry for QoreChain messages.

Building a transaction does not need a registry — a :class:`Msg`'s ``value`` is a
populated protobuf instance that packs directly into a Cosmos ``Any`` under its
``type_url`` (see :func:`qorechain.tx.send_messages`). The registry is the
*inverse*: it maps a ``type_url`` back to the generated class so a message pulled
off-chain (e.g. from a decoded ``TxBody``) can be parsed into a typed object,
and so callers can introspect the supported surface.

:data:`QORECHAIN_REGISTRY_TYPES` covers every QoreChain custom-module message;
the standard Cosmos SDK message classes cosmpy bundles are merged in by
:func:`qorechain_registry`. The ``type_url`` strings are the canonical on-chain
identifiers (``/qorechain.<module>.v1.Msg*`` and ``/cosmos.*``).
"""

from __future__ import annotations

from collections.abc import Callable
from typing import Any

from cosmpy.protos.cosmos.authz.v1beta1.tx_pb2 import MsgExec, MsgGrant, MsgRevoke
from cosmpy.protos.cosmos.bank.v1beta1.tx_pb2 import MsgMultiSend, MsgSend
from cosmpy.protos.cosmos.distribution.v1beta1.tx_pb2 import (
    MsgFundCommunityPool,
    MsgSetWithdrawAddress,
    MsgWithdrawDelegatorReward,
)
from cosmpy.protos.cosmos.feegrant.v1beta1.tx_pb2 import (
    MsgGrantAllowance,
    MsgRevokeAllowance,
)
from cosmpy.protos.cosmos.gov.v1.tx_pb2 import (
    MsgDeposit,
    MsgSubmitProposal,
    MsgVote,
    MsgVoteWeighted,
)
from cosmpy.protos.cosmos.staking.v1beta1.tx_pb2 import (
    MsgBeginRedelegate,
    MsgDelegate,
    MsgUndelegate,
)
from cosmpy.protos.ibc.applications.transfer.v1.tx_pb2 import MsgTransfer

from ..proto.qorechain.abstractaccount.v1 import tx_pb2 as abstractaccount_tx
from ..proto.qorechain.amm.v1 import tx_pb2 as amm_tx
from ..proto.qorechain.bridge.v1 import tx_pb2 as bridge_tx
from ..proto.qorechain.crossvm.v1 import tx_pb2 as crossvm_tx
from ..proto.qorechain.license.v1 import tx_pb2 as license_tx
from ..proto.qorechain.lightnode.v1 import tx_pb2 as lightnode_tx
from ..proto.qorechain.multilayer.v1 import tx_pb2 as multilayer_tx
from ..proto.qorechain.pqc.v1 import tx_pb2 as pqc_tx
from ..proto.qorechain.rdk.v1 import tx_pb2 as rdk_tx
from ..proto.qorechain.rlconsensus.v1 import tx_pb2 as rlconsensus_tx
from ..proto.qorechain.svm.v1 import tx_pb2 as svm_tx

#: A protobuf message class (callable producing a message instance).
ProtoType = Callable[..., Any]

#: Every QoreChain custom-module message, keyed by canonical on-chain type URL.
QORECHAIN_REGISTRY_TYPES: dict[str, ProtoType] = {
    # amm
    "/qorechain.amm.v1.MsgCreatePool": amm_tx.MsgCreatePool,
    "/qorechain.amm.v1.MsgAddLiquidity": amm_tx.MsgAddLiquidity,
    "/qorechain.amm.v1.MsgRemoveLiquidity": amm_tx.MsgRemoveLiquidity,
    "/qorechain.amm.v1.MsgSwapExactIn": amm_tx.MsgSwapExactIn,
    "/qorechain.amm.v1.MsgSwapExactOut": amm_tx.MsgSwapExactOut,
    "/qorechain.amm.v1.MsgPausePool": amm_tx.MsgPausePool,
    "/qorechain.amm.v1.MsgResumePool": amm_tx.MsgResumePool,
    # bridge
    "/qorechain.bridge.v1.MsgBridgeDeposit": bridge_tx.MsgBridgeDeposit,
    "/qorechain.bridge.v1.MsgBridgeWithdraw": bridge_tx.MsgBridgeWithdraw,
    "/qorechain.bridge.v1.MsgRegisterBridgeValidator": (
        bridge_tx.MsgRegisterBridgeValidator
    ),
    "/qorechain.bridge.v1.MsgBridgeAttestation": bridge_tx.MsgBridgeAttestation,
    # rdk
    "/qorechain.rdk.v1.MsgCreateRollup": rdk_tx.MsgCreateRollup,
    "/qorechain.rdk.v1.MsgSubmitBatch": rdk_tx.MsgSubmitBatch,
    "/qorechain.rdk.v1.MsgChallengeBatch": rdk_tx.MsgChallengeBatch,
    "/qorechain.rdk.v1.MsgResolveChallenge": rdk_tx.MsgResolveChallenge,
    "/qorechain.rdk.v1.MsgPauseRollup": rdk_tx.MsgPauseRollup,
    "/qorechain.rdk.v1.MsgResumeRollup": rdk_tx.MsgResumeRollup,
    "/qorechain.rdk.v1.MsgStopRollup": rdk_tx.MsgStopRollup,
    # multilayer
    "/qorechain.multilayer.v1.MsgRegisterSidechain": (
        multilayer_tx.MsgRegisterSidechain
    ),
    "/qorechain.multilayer.v1.MsgRegisterPaychain": multilayer_tx.MsgRegisterPaychain,
    "/qorechain.multilayer.v1.MsgAnchorState": multilayer_tx.MsgAnchorState,
    "/qorechain.multilayer.v1.MsgRouteTransaction": multilayer_tx.MsgRouteTransaction,
    "/qorechain.multilayer.v1.MsgUpdateLayerStatus": (
        multilayer_tx.MsgUpdateLayerStatus
    ),
    "/qorechain.multilayer.v1.MsgChallengeAnchor": multilayer_tx.MsgChallengeAnchor,
    # pqc
    "/qorechain.pqc.v1.MsgRegisterPQCKey": pqc_tx.MsgRegisterPQCKey,
    "/qorechain.pqc.v1.MsgRegisterPQCKeyV2": pqc_tx.MsgRegisterPQCKeyV2,
    "/qorechain.pqc.v1.MsgMigratePQCKey": pqc_tx.MsgMigratePQCKey,
    "/qorechain.pqc.v1.MsgDeprecateAlgorithm": pqc_tx.MsgDeprecateAlgorithm,
    "/qorechain.pqc.v1.MsgDisableAlgorithm": pqc_tx.MsgDisableAlgorithm,
    # svm
    "/qorechain.svm.v1.MsgDeployProgram": svm_tx.MsgDeployProgram,
    "/qorechain.svm.v1.MsgCreateAccount": svm_tx.MsgCreateAccount,
    "/qorechain.svm.v1.MsgExecuteProgram": svm_tx.MsgExecuteProgram,
    "/qorechain.svm.v1.MsgRegisterSVMPQCKey": svm_tx.MsgRegisterSVMPQCKey,
    # lightnode
    "/qorechain.lightnode.v1.MsgRegisterLightNode": (
        lightnode_tx.MsgRegisterLightNode
    ),
    "/qorechain.lightnode.v1.MsgHeartbeat": lightnode_tx.MsgHeartbeat,
    "/qorechain.lightnode.v1.MsgDeregisterLightNode": (
        lightnode_tx.MsgDeregisterLightNode
    ),
    "/qorechain.lightnode.v1.MsgClaimLightNodeRewards": (
        lightnode_tx.MsgClaimLightNodeRewards
    ),
    # license
    "/qorechain.license.v1.MsgGrantLicense": license_tx.MsgGrantLicense,
    "/qorechain.license.v1.MsgRevokeLicense": license_tx.MsgRevokeLicense,
    "/qorechain.license.v1.MsgSuspendLicense": license_tx.MsgSuspendLicense,
    "/qorechain.license.v1.MsgResumeLicense": license_tx.MsgResumeLicense,
    # abstractaccount
    "/qorechain.abstractaccount.v1.MsgCreateAbstractAccount": (
        abstractaccount_tx.MsgCreateAbstractAccount
    ),
    "/qorechain.abstractaccount.v1.MsgUpdateSpendingRules": (
        abstractaccount_tx.MsgUpdateSpendingRules
    ),
    # crossvm
    "/qorechain.crossvm.v1.MsgCrossVMCall": crossvm_tx.MsgCrossVMCall,
    "/qorechain.crossvm.v1.MsgProcessQueue": crossvm_tx.MsgProcessQueue,
    # rlconsensus
    "/qorechain.rlconsensus.v1.MsgSetAgentMode": rlconsensus_tx.MsgSetAgentMode,
    "/qorechain.rlconsensus.v1.MsgResumeAgent": rlconsensus_tx.MsgResumeAgent,
    "/qorechain.rlconsensus.v1.MsgUpdatePolicy": rlconsensus_tx.MsgUpdatePolicy,
    "/qorechain.rlconsensus.v1.MsgUpdateRewardWeights": (
        rlconsensus_tx.MsgUpdateRewardWeights
    ),
}

#: The standard Cosmos SDK + IBC messages this SDK composes, keyed by type URL.
COSMOS_REGISTRY_TYPES: dict[str, ProtoType] = {
    "/cosmos.bank.v1beta1.MsgSend": MsgSend,
    "/cosmos.bank.v1beta1.MsgMultiSend": MsgMultiSend,
    "/cosmos.staking.v1beta1.MsgDelegate": MsgDelegate,
    "/cosmos.staking.v1beta1.MsgUndelegate": MsgUndelegate,
    "/cosmos.staking.v1beta1.MsgBeginRedelegate": MsgBeginRedelegate,
    "/cosmos.distribution.v1beta1.MsgWithdrawDelegatorReward": (
        MsgWithdrawDelegatorReward
    ),
    "/cosmos.distribution.v1beta1.MsgSetWithdrawAddress": MsgSetWithdrawAddress,
    "/cosmos.distribution.v1beta1.MsgFundCommunityPool": MsgFundCommunityPool,
    "/cosmos.gov.v1.MsgVote": MsgVote,
    "/cosmos.gov.v1.MsgVoteWeighted": MsgVoteWeighted,
    "/cosmos.gov.v1.MsgDeposit": MsgDeposit,
    "/cosmos.gov.v1.MsgSubmitProposal": MsgSubmitProposal,
    "/cosmos.authz.v1beta1.MsgGrant": MsgGrant,
    "/cosmos.authz.v1beta1.MsgRevoke": MsgRevoke,
    "/cosmos.authz.v1beta1.MsgExec": MsgExec,
    "/cosmos.feegrant.v1beta1.MsgGrantAllowance": MsgGrantAllowance,
    "/cosmos.feegrant.v1beta1.MsgRevokeAllowance": MsgRevokeAllowance,
    "/ibc.applications.transfer.v1.MsgTransfer": MsgTransfer,
}


def qorechain_registry(
    extra_types: dict[str, ProtoType] | None = None,
) -> dict[str, ProtoType]:
    """Build the full type-URL → protobuf-class map.

    Merges the standard Cosmos SDK / IBC messages and every QoreChain custom
    message. ``extra_types`` (e.g. private-module messages) is applied last and
    overrides any colliding built-in entry.

    :param extra_types: Optional additional ``{type_url: ProtoType}`` entries.
    :returns: A fresh dict the caller may mutate freely.
    """
    registry: dict[str, ProtoType] = {**COSMOS_REGISTRY_TYPES, **QORECHAIN_REGISTRY_TYPES}
    if extra_types:
        registry.update(extra_types)
    return registry


def resolve_message_type(
    type_url: str, registry: dict[str, ProtoType] | None = None
) -> ProtoType:
    """Look up the protobuf class for a ``type_url``.

    :param type_url: A canonical ``/package.Msg*`` identifier.
    :param registry: An optional registry; defaults to :func:`qorechain_registry`.
    :raises KeyError: If the type URL is not registered.
    """
    reg = registry if registry is not None else qorechain_registry()
    return reg[type_url]


def decode_any(
    type_url: str, value: bytes, registry: dict[str, ProtoType] | None = None
) -> Any:
    """Decode raw ``Any`` bytes into a typed protobuf message.

    :param type_url: The ``Any.type_url`` identifying the message.
    :param value: The ``Any.value`` (serialized message bytes).
    :param registry: An optional registry; defaults to :func:`qorechain_registry`.
    :raises KeyError: If the type URL is not registered.
    """
    msg = resolve_message_type(type_url, registry)()
    msg.ParseFromString(value)
    return msg

File: packages/go/qorechain/messages/registry.go
// Package messages provides the interface registry, codec, and typed message
// composers for every transaction QoreChain supports.
//
// The registry registers all 53 custom QoreChain Msg implementations (across the
// 11 custom modules) plus the standard Cosmos SDK modules under their type URLs,
// so a custom Msg can be packed into a tx Any and decoded back through the codec
// exactly like the chain does. RegisterInterfaces mirrors each module's
// RegisterInterfaces on the chain.
package messages

import (
	feegranttypes "cosmossdk.io/x/feegrant"
	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
	cryptocodec "github.com/cosmos/cosmos-sdk/crypto/codec"
	sdk "github.com/cosmos/cosmos-sdk/types"
	txtypes "github.com/cosmos/cosmos-sdk/types/tx"
	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
	authztypes "github.com/cosmos/cosmos-sdk/x/authz"
	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
	distrtypes "github.com/cosmos/cosmos-sdk/x/distribution/types"
	govv1types "github.com/cosmos/cosmos-sdk/x/gov/types/v1"
	govv1beta1types "github.com/cosmos/cosmos-sdk/x/gov/types/v1beta1"
	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
	ibctransfertypes "github.com/cosmos/ibc-go/v8/modules/apps/transfer/types"

	abstractaccountv1 "github.com/qorechain/qorechain-sdk/packages/go/qorechain/proto/qorechain/abstractaccount/v1"
	ammv1 "github.com/qorechain/qorechain-sdk/packages/go/qorechain/proto/qorechain/amm/v1"
	bridgev1 "github.com/qorechain/qorechain-sdk/packages/go/qorechain/proto/qorechain/bridge/v1"
	crossvmv1 "github.com/qorechain/qorechain-sdk/packages/go/qorechain/proto/qorechain/crossvm/v1"
	licensev1 "github.com/qorechain/qorechain-sdk/packages/go/qorechain/proto/qorechain/license/v1"
	lightnodev1 "github.com/qorechain/qorechain-sdk/packages/go/qorechain/proto/qorechain/lightnode/v1"
	multilayerv1 "github.com/qorechain/qorechain-sdk/packages/go/qorechain/proto/qorechain/multilayer/v1"
	pqcv1 "github.com/qorechain/qorechain-sdk/packages/go/qorechain/proto/qorechain/pqc/v1"
	rdkv1 "github.com/qorechain/qorechain-sdk/packages/go/qorechain/proto/qorechain/rdk/v1"
	rlconsensusv1 "github.com/qorechain/qorechain-sdk/packages/go/qorechain/proto/qorechain/rlconsensus/v1"
	svmv1 "github.com/qorechain/qorechain-sdk/packages/go/qorechain/proto/qorechain/svm/v1"
)

// RegisterInterfaces registers every QoreChain custom Msg implementation plus the
// standard Cosmos SDK message and crypto interfaces into reg. After this call,
// the codec can pack any of these messages into an Any and unpack them back.
func RegisterInterfaces(reg codectypes.InterfaceRegistry) {
	registerQoreChainInterfaces(reg)
	registerCosmosInterfaces(reg)
}

// registerQoreChainInterfaces registers all 53 custom QoreChain Msg
// implementations under the sdk.Msg interface.
func registerQoreChainInterfaces(reg codectypes.InterfaceRegistry) {
	reg.RegisterImplementations((*sdk.Msg)(nil),
		// amm (7)
		&ammv1.MsgCreatePool{},
		&ammv1.MsgAddLiquidity{},
		&ammv1.MsgRemoveLiquidity{},
		&ammv1.MsgSwapExactIn{},
		&ammv1.MsgSwapExactOut{},
		&ammv1.MsgPausePool{},
		&ammv1.MsgResumePool{},
		// bridge (7)
		&bridgev1.MsgBridgeDeposit{},
		&bridgev1.MsgBridgeWithdraw{},
		&bridgev1.MsgRegisterBridgeValidator{},
		&bridgev1.MsgBridgeAttestation{},
		&bridgev1.MsgUpdateEthLightClient{},
		&bridgev1.MsgUpdateChainConfig{},
		&bridgev1.MsgSetVerifierBootstrap{},
		// rdk (8)
		&rdkv1.MsgCreateRollup{},
		&rdkv1.MsgSubmitBatch{},
		&rdkv1.MsgChallengeBatch{},
		&rdkv1.MsgResolveChallenge{},
		&rdkv1.MsgPauseRollup{},
		&rdkv1.MsgResumeRollup{},
		&rdkv1.MsgStopRollup{},
		&rdkv1.MsgExecuteWithdrawal{},
		// multilayer (6)
		&multilayerv1.MsgRegisterSidechain{},
		&multilayerv1.MsgRegisterPaychain{},
		&multilayerv1.MsgAnchorState{},
		&multilayerv1.MsgRouteTransaction{},
		&multilayerv1.MsgUpdateLayerStatus{},
		&multilayerv1.MsgChallengeAnchor{},
		// pqc (5)
		&pqcv1.MsgRegisterPQCKey{},
		&pqcv1.MsgRegisterPQCKeyV2{},
		&pqcv1.MsgMigratePQCKey{},
		&pqcv1.MsgDeprecateAlgorithm{},
		&pqcv1.MsgDisableAlgorithm{},
		// svm (4)
		&svmv1.MsgDeployProgram{},
		&svmv1.MsgCreateAccount{},
		&svmv1.MsgExecuteProgram{},
		&svmv1.MsgRegisterSVMPQCKey{},
		// lightnode (4)
		&lightnodev1.MsgRegisterLightNode{},
		&lightnodev1.MsgHeartbeat{},
		&lightnodev1.MsgDeregisterLightNode{},
		&lightnodev1.MsgClaimLightNodeRewards{},
		// license (4)
		&licensev1.MsgGrantLicense{},
		&licensev1.MsgRevokeLicense{},
		&licensev1.MsgSuspendLicense{},
		&licensev1.MsgResumeLicense{},
		// abstractaccount (2)
		&abstractaccountv1.MsgCreateAbstractAccount{},
		&abstractaccountv1.MsgUpdateSpendingRules{},
		// crossvm (2)
		&crossvmv1.MsgCrossVMCall{},
		&crossvmv1.MsgProcessQueue{},
		// rlconsensus (4)
		&rlconsensusv1.MsgSetAgentMode{},
		&rlconsensusv1.MsgResumeAgent{},
		&rlconsensusv1.MsgUpdatePolicy{},
		&rlconsensusv1.MsgUpdateRewardWeights{},
	)
}

// registerCosmosInterfaces registers the standard Cosmos SDK message, account,
// and crypto interfaces so the SDK can build bank/staking/distribution/gov/
// authz/feegrant transactions and resolve public keys in tx AuthInfo.
func registerCosmosInterfaces(reg codectypes.InterfaceRegistry) {
	cryptocodec.RegisterInterfaces(reg)
	authtypes.RegisterInterfaces(reg)
	banktypes.RegisterInterfaces(reg)
	stakingtypes.RegisterInterfaces(reg)
	distrtypes.RegisterInterfaces(reg)
	govv1types.RegisterInterfaces(reg)
	govv1beta1types.RegisterInterfaces(reg)
	authztypes.RegisterInterfaces(reg)
	feegranttypes.RegisterInterfaces(reg)
	ibctransfertypes.RegisterInterfaces(reg)
	// Register the tx message-service descriptors so msgservice resolution works.
	txtypes.RegisterInterfaces(reg)
}
